File: tank_player_templates/tank_player_java/src/cwg/Bfs.java
/******************************************************************************
*
* Copyright Consoden AB, 2014
*
* Created by: Joel Ottosson / joot
*
******************************************************************************/
package cwg;

//Implementation of a breadth first algorithm
//Helpful to find out if a square is reachable and get the shortest path.
class Bfs {
	
	private consoden.tankgame.GameState gameState;	
	private Position startPos;
	private int[][] gamePaths = null;
	
	private byte empty=(byte)'.';
	private byte coin=(byte)'$';
	private byte poison=(byte)'p';
	
	//Create a bfs graph based on startPos.
	public Bfs(consoden.tankgame.GameState gameState, Position startPos) {		
		this.gameState=gameState;
		this.startPos=startPos;
		gamePaths=new int[getSizeX()][getSizeY()];
		for (int x=0; x<getSizeX(); x++) {
			for (int y=0; y<getSizeY(); y++) {
				gamePaths[x][y] = Integer.MAX_VALUE;
			}
		}

		gamePaths[startPos.x][startPos.y] = 0;
					
		evaluateShortestPaths (startPos, 1);
	}
	
	//Check if square is reachable.
	public boolean canReachSquare(Position pos) {
		if (pos.x >= getSizeX() || pos.x < 0 || pos.y >= getSizeY() || pos.y < 0) {
			return false;
		}
		return gamePaths[pos.x][pos.y] < Integer.MAX_VALUE;
	}

	//Least number of moves needed to reach square.
	public int stepsToSquare(Position pos) {
		if (pos.x >= getSizeX() || pos.x < 0 || pos.y >= getSizeY() || pos.y < 0) {
			return Integer.MAX_VALUE;
		}
		return gamePaths[pos.x][pos.y];
	}
	
	//Find the direction to move for shortest path to target.
	public consoden.tankgame.Direction backtrackFromSquare(Position target) {
		Position nextMove = backtrack (target);

		// Shortest path is one step less
		if (nextMove.equals(move(startPos, consoden.tankgame.Direction.LEFT)))
			return consoden.tankgame.Direction.LEFT;
		else if (nextMove.equals(move(startPos, consoden.tankgame.Direction.RIGHT)))
			return consoden.tankgame.Direction.RIGHT;
		else if (nextMove.equals(move(startPos, consoden.tankgame.Direction.UP)))
			return consoden.tankgame.Direction.UP;
		else if (nextMove.equals(move(startPos, consoden.tankgame.Direction.DOWN)))
			return consoden.tankgame.Direction.DOWN;
		else
			return consoden.tankgame.Direction.NEUTRAL;
	}
	
	private Position backtrack(Position target) {
		if (!canReachSquare(target)) {
			return new Position (-1, -1);
		}

		// Steps to square
		int steps = stepsToSquare(target);

		// one step - found the next square
		if (steps == 1) {
			return target;
		}

		// Shortest path is one step less
		Position left = move(target, consoden.tankgame.Direction.LEFT);
		if (stepsToSquare(left) == (steps-1)) {
			return backtrack(left);
		}

		Position right = move(target, consoden.tankgame.Direction.RIGHT);
		if (stepsToSquare(right) == (steps-1)) {
			return backtrack(right);
		}

		Position up = move(target, consoden.tankgame.Direction.UP);
		if (stepsToSquare(up) == (steps-1)) {
			return backtrack(up);
		}

		Position down = move(target, consoden.tankgame.Direction.DOWN);
		if (stepsToSquare(down) == (steps-1)) {
			return backtrack(down);
		}

		// How did this happend??
		return new Position (-1, -1);	
	}
	
	//X size of game field
	private int getSizeX() {
		return gameState.width().getVal();
	}
	
	//Y size of game field
	private int getSizeY() {
		return gameState.height().getVal();
	}
		
	private Position move(final Position p, final consoden.tankgame.Direction d) {
		switch (d) {
        case LEFT:
        	return new Position((p.x-1+getSizeX())%getSizeX(), p.y);
        case RIGHT:
        	return new Position((p.x+1)%getSizeX(), p.y);
        case UP:
        	return new Position(p.x, (p.y-1+getSizeY())%getSizeY());
        case DOWN:
        	return new Position(p.x, (p.y+1)%getSizeY());
        case NEUTRAL:
        	return new Position(p.x, p.y);
        }
		return new Position(p.x, p.y);
	}

	private int toIndex(int x, int y) {
		return x+y*getSizeX();
	}
	
	private byte rawVal(int x, int y) {
		return gameState.board().getVal()[toIndex(x, y)];
	}
	
	private void evaluateShortestPathsHelper(Position nextPos, int steps) {		
		byte next = rawVal(nextPos.x, nextPos.y);
		boolean isEmpty= next==empty || next==coin || next==poison;
		if (isEmpty) 
		{
			if (steps < gamePaths[nextPos.x][nextPos.y]) {
				System.out.println(nextPos.toString()+" evaluated step: "+steps);
				gamePaths[nextPos.x][nextPos.y] = steps;
				evaluateShortestPaths(nextPos, steps+1);                
			}
		}
		else
		{
			System.out.println(nextPos.toString()+" not empty: "+(char)next);
		}
	}

	private void evaluateShortestPaths(Position pos, int steps) {
		evaluateShortestPathsHelper(move(pos, consoden.tankgame.Direction.LEFT), steps);
		evaluateShortestPathsHelper(move(pos, consoden.tankgame.Direction.RIGHT), steps);
		evaluateShortestPathsHelper(move(pos, consoden.tankgame.Direction.UP), steps);
		evaluateShortestPathsHelper(move(pos, consoden.tankgame.Direction.DOWN), steps);
	}
	
	/*private void dump() {
		System.out.println("--- path ---");
		for (int y=0; y<getSizeY(); y++) {
			for (int x=0; x<getSizeX(); x++) {
				if (gamePaths[x][y]<Integer.MAX_VALUE)
					System.out.print(gamePaths[x][y]);
				else
					System.out.print("X");
				System.out.print(" ");
			}
			System.out.println();
		}
	}*/
}

File: src/players/tank_player_templates/tank_player_cs/src/Bfs.cs
/******************************************************************************
*
* Copyright Consoden AB, 2014
*
* Created by: Joel Ottosson / joot
*
******************************************************************************/
using System;

namespace tank_player_cs
{
	//Implementation of a breadth first algorithm
	//Helpful to find out if a square is reachable and get the shortest path.
	class Bfs
	{
		private Consoden.TankGame.GameState gameState;
		private int[,] gamePaths = null;
		private Position startPos;
		private bool allowGas = true;

		//Create a bfs graph based on startPos.
		public Bfs (Consoden.TankGame.GameState gameState, Position startPos, bool allowGas = true)
		{
			this.gameState=gameState;
			this.startPos = startPos;
			this.allowGas = allowGas;

			gamePaths = new int[SizeX, SizeY];
			for (int x=0; x<SizeX; x++) {
				for (int y=0; y<SizeY; y++) {
					gamePaths [x, y] = int.MaxValue;
				}
			}

			gamePaths [startPos.X, startPos.Y] = 0;
			EvaluateShortestPaths (startPos, 1);
		}	

		//Check if square is reachable.
		public bool CanReachSquare(Position pos)
		{
			if (pos.X >= SizeX || pos.X < 0 || pos.Y >= SizeY || pos.Y < 0) {
				return false;
			}
			return gamePaths[pos.X, pos.Y] < int.MaxValue;
		}

		//Least number of moves needed to reach square.
		public int StepsToSquare(Position pos)
		{
			if (pos.X >= SizeX || pos.X < 0 || pos.Y >= SizeY || pos.Y < 0) {
				return int.MaxValue;
			}
			return gamePaths[pos.X, pos.Y];
		}

		//Find the direction to move for shortest path to target.
		public Consoden.TankGame.Direction.Enumeration BacktrackFromSquare(Position target)
		{
			var nextMove = Backtrack (target);

			// Shortest path is one step less
			if (Move (startPos, Consoden.TankGame.Direction.Enumeration.Left) == nextMove)
				return Consoden.TankGame.Direction.Enumeration.Left;
			else if (Move (startPos, Consoden.TankGame.Direction.Enumeration.Right) == nextMove)
				return Consoden.TankGame.Direction.Enumeration.Right;
			else if (Move (startPos, Consoden.TankGame.Direction.Enumeration.Up) == nextMove)
				return Consoden.TankGame.Direction.Enumeration.Up;
			else if (Move (startPos, Consoden.TankGame.Direction.Enumeration.Down) == nextMove)
				return Consoden.TankGame.Direction.Enumeration.Down;
			else
				return Consoden.TankGame.Direction.Enumeration.Neutral;
		}


		private Position Backtrack(Position target)
		{
			if (!CanReachSquare(target)) {
				return new Position (-1, -1);
			}

			// Steps to square
			int steps = StepsToSquare(target);

			// one step - found the next square
			if (steps == 1) {
				return target;
			}

			// Shortest path is one step less
			var left = Move (target, Consoden.TankGame.Direction.Enumeration.Left);
			if (StepsToSquare(left) == (steps-1)) {
				return Backtrack(left);
			}

			var right = Move (target, Consoden.TankGame.Direction.Enumeration.Right);
			if (StepsToSquare(right) == (steps-1)) {
				return Backtrack(right);
			}

			var up = Move (target, Consoden.TankGame.Direction.Enumeration.Up);
			if (StepsToSquare(up) == (steps-1)) {
				return Backtrack(up);
			}

			var down = Move (target, Consoden.TankGame.Direction.Enumeration.Down);
			if (StepsToSquare(down) == (steps-1)) {
				return Backtrack(down);
			}

			// How did this happend??
			return new Position (-1, -1);	
		}

		private void EvaluateShortestPathsHelper(Position nextPos, int steps) 
		{
			char next = RawVal (nextPos.X, nextPos.Y);
			bool isMine = next == 'o';
			bool isWall = next == 'x';
			bool isPoison = next == 'p';

			if (!(isMine || isWall) && (allowGas || !isPoison)) 
			{
				if (steps < gamePaths[nextPos.X, nextPos.Y]) {
					gamePaths[nextPos.X, nextPos.Y] = steps;
					EvaluateShortestPaths(nextPos, steps+1);                
				}
			}
		}

		private void EvaluateShortestPaths(Position pos, int steps)
		{
			EvaluateShortestPathsHelper(Move(pos, Consoden.TankGame.Direction.Enumeration.Left), steps);
			EvaluateShortestPathsHelper(Move(pos, Consoden.TankGame.Direction.Enumeration.Right), steps);
			EvaluateShortestPathsHelper(Move(pos, Consoden.TankGame.Direction.Enumeration.Up), steps);
			EvaluateShortestPathsHelper(Move(pos, Consoden.TankGame.Direction.Enumeration.Down), steps);
		}

		//X size of game field
		private int SizeX {
			get {
				return gameState.Width.Val;
			}
		}

		//Y size of game field
		private int SizeY {
			get {
				return gameState.Height.Val;
			}
		}

		private int ToIndex (int x, int y)
		{
			return x + y * SizeX;
		}

		private char RawVal(int x, int y)
		{
			int i = ToIndex (x, y);
			return Convert.ToChar (gameState.Board.Val [i]);
		}

		//Move p one step in specified direction and returns the new position.
		public Position Move(Position p, Consoden.TankGame.Direction.Enumeration d)
		{
			switch (d) {
				case Consoden.TankGame.Direction.Enumeration.Left:
				return new Position ((p.X - 1 + SizeX) % SizeX, p.Y);
				case Consoden.TankGame.Direction.Enumeration.Right:
				return new Position ((p.X + 1) % SizeX, p.Y);
				case Consoden.TankGame.Direction.Enumeration.Up:
				return new Position (p.X, (p.Y - 1 + SizeY) % SizeY);
				case Consoden.TankGame.Direction.Enumeration.Down:
				return new Position (p.X, (p.Y + 1) % SizeY);
				default:
				return p;
			}
		}
	}
}
